r2r_tracing/src/rclcpp_tracepoints.rs

```diff
@@ -0,0 +1,156 @@
+            subscription_handle.cast(),
+            ref_to_c_void(subscription),
+        );
+    }
+}
+
+/// Link a subscription callback identified by `callback_id` to a `subscription` object.
+///
+/// Tracepoint: `ros2::rclcpp_subscription_callback_added`
+pub fn trace_subscription_callback_added<S>(subscription: &S, callback_id: usize) {
+    unsafe {
+        tp::ros_trace_rclcpp_subscription_callback_added(
+            ref_to_c_void(subscription),
+            c_void!(callback_id),
+        );
+    }
+}
+
+/// Message taking.
+///
+/// Notes the **reference** to the `message` being taken at the `rclcpp`/`r2r` level.
+///
+/// To trace void pointer to the message use [`trace_take_ptr`].
+///
+/// Tracepoint: `ros2::rclcpp_take`
+pub fn trace_take<M>(message: &M) {
+    unsafe {
+        tp::ros_trace_rclcpp_take(ref_to_c_void(message));
+    }
+}
+
+/// Message taking.
+///
+/// Notes the **void pointer** to the `message` being taken at the `rclcpp`/`r2r` level.
+/// The message pointer is NOT dereferenced.
+///
+/// To trace reference to the message use [`trace_take`].
+///
+/// Tracepoint: `ros2::rclcpp_take`
+#[allow(clippy::not_unsafe_ptr_arg_deref)]
+pub fn trace_take_ptr(message: *const std::ffi::c_void) {
+    unsafe {
+        tp::ros_trace_rclcpp_take(message);
+    }
+}
+
+/// Link a service callback identified by `callback_id` to a `service`.
+///
+/// Tracepoint: `ros2::rclcpp_service_callback_added`
+pub fn trace_service_callback_added(service: *const rcl_service_t, callback_id: usize) {
+    unsafe { tp::ros_trace_rclcpp_service_callback_added(service.cast(), c_void!(callback_id)) }
+}
+
+/// Link a timer callback identified by `callback_id` to its `rcl_timer_t` handle.
+///
+/// Tracepoint: `ros2::rclcpp_timer_callback_added`
+pub fn trace_timer_callback_added(timer: *const rcl_timer_t, callback_id: usize) {
+    unsafe {
+        tp::ros_trace_rclcpp_timer_callback_added(timer.cast(), c_void!(callback_id));
+    }
+}
+
+/// Link a `timer` to a `node`.
+///
+/// Tracepoint: `ros2::rclcpp_timer_link_node`
+pub fn trace_timer_link_node(timer: *const rcl_timer_t, node: *const rcl_node_t) {
+    unsafe {
+        tp::ros_trace_rclcpp_timer_link_node(timer.cast(), node.cast());
+    }
+}
+
+/// Register a demangled `function_symbol` with a `callback_id`.
+///
+/// Allocates memory to store the function symbol as a `CString`.
+///
+/// Tracepoint: `ros2::callback_register`
+///
+/// # Panics
+/// If `function_symbol` contains a null byte.
+pub fn trace_callback_register(callback_id: usize, function_symbol: &str) {
+    let function_symbol = CString::new(function_symbol)
+        .expect("r2r tracing: Cannot convert function_symbol to CString. It contains null byte.");
+
+    unsafe {
+        tp::ros_trace_rclcpp_callback_register(c_void!(callback_id), function_symbol.as_ptr());
+    }
+}
+
+/// Start of a callback
+///
+/// If this callback is done via intra-process set `is_intra_process` to `true`.
```

Comment:
///  Set `is_intra_process` depending on whether this callback is done via intra-process or not